include/swift/SIL/SILValue.h

```diff
@@ -128,24 +130,45 @@ struct ValueOwnershipKind {
   bool isTrivialOr(ValueOwnershipKind Kind) const {
     return Value == Trivial || Value == Kind;
   }
+
+  /// Given that there is an aggregate value (like a struct or enum) with this
+  /// ownership kind, and a subobject of type Proj is being projected from the
+  /// aggregate, return Trivial if Proj has trivial type and the aggregate's
+  /// ownership kind otherwise.
+  ValueOwnershipKind getProjectedOwnershipKind(SILModule &M,
+                                               SILType Proj) const;
 };
 
 llvm::raw_ostream &operator<<(llvm::raw_ostream &os, ValueOwnershipKind Kind);
 
 /// This is the base class of the SIL value hierarchy, which represents a
-/// runtime computed value. Things like SILInstruction derive from this.
+/// runtime computed value. Some examples of ValueBase are SILArgument and
+/// SingleValueInstruction.
 class ValueBase : public SILNode, public SILAllocated<ValueBase> {
+  friend class Operand;
+
   SILType Type;
   Operand *FirstUse = nullptr;
-  friend class Operand;
+
+  /// A 32 bit field of data that we provide for abstract subclasses to use.
+  uint32_t SubclassData;
 
   ValueBase(const ValueBase &) = delete;
   ValueBase &operator=(const ValueBase &) = delete;
 
 protected:
-  ValueBase(ValueKind kind, SILType type)
-    : SILNode(SILNodeKind(kind), SILNodeStorageLocation::Value),
-      Type(type) {}
+  ValueBase(ValueKind kind, SILType type, SILNodeIsCanonical IsCanonical)
+      : SILNode(SILNodeKind(kind), SILNodeStorageLocation::Value, IsCanonical),
+        Type(type), SubclassData(0) {}
+
+  ArrayRef<uint8_t> getSubclassData() const {
+    return {reinterpret_cast<const uint8_t *>(&SubclassData),
+            sizeof(SubclassData)};
+  }
+
+  MutableArrayRef<uint8_t> getSubclassData() {
+    return {reinterpret_cast<uint8_t *>(&SubclassData), sizeof(SubclassData)};
+  }
```

Comment:
This seems unnecessary.

Reply:
I was just trying to provide a nice wrapper. I can't do this anyways, once I pack into a bitfield in SILNode.